Dsl feature

ZenGround0 opened the pull request:
The DSL selection feature chooses nodes on which to run commands.  Below is the original specification I proposed for the feature.  This pull request includes changes to the kubernetes-ipfs executable that implement node selection, and a set of tests that can be used both to sanity check that this feature is implemented correctly and as a reference for the different functionality of the node selection feature.

Note that this PR includes commits from David, Frank, and Hector in the history.  I'm happy to merge to a different branch or do some kind of history cleanup if this is desired.

The proposed spec:
   Proposed addition to kIPFS DSL to generalize selection of nodes 
   on which to run commands.  

   Overview
   ************************************
   Instead of selecting nodes with 
   OnNode, EndNode, each step takes a selection object which allows
   tests to specify nodes in 3 ways

   1: Select nodes by range like with OnNode EndNode
      - Can choose OnNode EndNode
      - Can choose Number, and Number of random nodes will run
      **Example**
```
      selection:
        range:
          order: SEQUENTIAL
          start: 1
          end: 10
```


   2: Select percentage of nodes to run
      - Can choose Start, Percentage (run 30 % of nodes starting at node 2)
      - Can choose Percentage (run 30 % of nodes choosing at random)
      **Example**
```
      selection:
        percent: 
          order: RANDOM
          percent: 33
```

   3: Select a subset of nodes to run
      - Choose "subset indices" which are specified in the configuration
      - One partition of nodes into subsets per test (config)
      **Example**
```
      selection:
        subset:
          indices: [4, 7, 8]
```


   Configuration:
   ************************************
   To specify the partition of nodes into subsets for a given test, we use the
   config structure.  The first choice to make is whether to divide the nodes
   up evenly (using David's current subset implementation), or into specified
   chunks by percentage.  The next choice is whether nodes should fill 
   partitions sequentially, or be selected at random
   **Example**
   config:
```
   ...
     subset_partion:
       partition_type: WEIGHTED
       order: random
       percents: [10, 5, 85]
   ...
```

   **Example2**
```
   config:
   ...
     subset_partion:
       partition_type: EVEN
       order: sequential
       number_partitions: 8
   ...

```

   Backwards Compatability:
   ************************************
   As described this will break backwards compatability.  Although we could go
   back and change existing tests there is something to be said for the 
   simplicity of OnNode: 1 as opposed to 
   ```
    selection: 
         range:
          order: SEQUENTIAL
          start: 1
          end: 1

```
   For this reason this feature still supports simple onNode/EndNode functionality alongside Selections

hsanjuan commented on ipfs-cluster/test_tests/README.txt:
```diff
@@ -0,0 +1,9 @@
+These tests are used for sanity checking the DSL implementation of node
```
You may want to move these files up one level, as it's not ipfs-cluster specific.

hsanjuan commented on ipfs-cluster/test_tests/old_range.yml:
```diff
@@ -0,0 +1,14 @@
+  - name: Run on old step nodes 2 3 4
+    cmd: echo "old range" > OLD_RANGE
+    on_node: 2
+    end_node: 4
```
General comments about the tests:

It seems you rely on information echo-ed to the screen to see that things work and make sense.

Instead you should find a way to verify that the test worked with `assertions`. Maybe the `assertions` part of the DSL does not support verifying that something happened on, say, 1/3 of nodes. Then we probably need to extend that part of the language too. It does not necessarily need to happen in this PR though.

ZenGround0 replied:
Thanks Hector, I realize that this is a hack.  As you said, I could not find a way to specifically identify which node was being called using assertions in the DSL.  

It turned out that the echoed output was actually less important than looking at the testing output.  I tested once to see that the indices passed in to handleStep are actually getting run, and then for the rest of the tests the real check is to see whether the expected nodes (in the test name) are the same as the actual nodes (output by default by the test framework).  Maybe it would be cleaner to make the command a noop of some kind?

hsanjuan commented on main.go:
```diff
@@ -48,23 +56,71 @@ type Assertion struct {
+	WriteToFile string      `yaml:"write_to_file"`
+}
+
+// Selection is
```
`Selection is` what ? :) Would be nice to put a small description and the examples from the proposal here (and in the next struct definitions). Note that this get's rendered by Godoc later...

ZenGround0 replied:
Ah got it, that makes sense.  Will update

hsanjuan commented on main.go:
```diff
@@ -396,6 +486,177 @@ func runInPod(name string, cmdToRun string, env []string, timeout int) ([]string
+		}
+		return makeRange(step.OnNode, step.EndNode)
+	} else if step.Selection == nil {
+		fatal("No selection method on test step")
```
I think the cleanest thing here split: into multiple functions:

```
switch (step) {
case step.OnNode <=0 && step.selection == nil:
   fail
case step.OnNode > 0 && step.selection != nil:
  fail
case stepOnNode > 0:
  selectNodesFromOn()
case step.Selection != nil:
  selectNodesFromSelection()
default:
  fail
```

ZenGround0 replied:
I like this better too.  One unfortunate thing is that here we can't switch on step so I am going with the slightly if else variant of this.

ZenGround0 replied:
Ah I think I missed some of your later comments.  I will refactor in favor of a selection validation function tomorrow.

hsanjuan commented on main.go:
```diff
@@ -396,6 +486,177 @@ func runInPod(name string, cmdToRun string, env []string, timeout int) ([]string
+	} else if step.Selection == nil {
+		fatal("No selection method on test step")
+	} else { /* Selection chooses nodes */
+		if (step.Selection.Range == nil && step.Selection.Percent == nil && step.Selection.Subset == nil) {
```
Maybe a switch statement?

hsanjuan commented on main.go:
```diff
@@ -396,6 +486,177 @@ func runInPod(name string, cmdToRun string, env []string, timeout int) ([]string
+
+func selectNodesRange(step *Step, config Config) []int{
+	var selection []int
+	if step.Selection.Range.Order == SEQUENTIAL {
```
switch statement

hsanjuan commented on main.go:
```diff
@@ -396,6 +486,177 @@ func runInPod(name string, cmdToRun string, env []string, timeout int) ([]string
+func selectNodesRange(step *Step, config Config) []int{
+	var selection []int
+	if step.Selection.Range.Order == SEQUENTIAL {
+	  if step.Selection.Range.Start <= 0 || 
```
I have the feeling you need a "ValidateSelections()" function which runs before running any tests to make sure things make sense.

Exiting half-way through the tests because of an obvious problem in the selection parameters is not nice. Once you have that you can select your nodes in a much clearer way without this if-then-else nesting.

Also the Go-way tends to bubble up errors, rather than calling `fatal` all around. For example, deciding that these errors are not a fatal and the tests should continue after informing the user needs changing a lot of lines. If these errors were bubbled, it would be matter of changing just the place where `selectNodes` is called.

ZenGround0 replied:
@hsanjuan Noted, I will refactor into pre-validation and bubble up errors in the next changes.

hsanjuan commented on main.go:
```diff
@@ -396,6 +486,177 @@ func runInPod(name string, cmdToRun string, env []string, timeout int) ([]string
+		fatal("Invalid percent")
+	}
+	numNodes := int( (float64(percent) / 100.0) * float64(config.Nodes) )
+	if step.Selection.Percent.Order == SEQUENTIAL {
```
switch

hsanjuan commented on main.go:
```diff
@@ -396,6 +486,177 @@ func runInPod(name string, cmdToRun string, env []string, timeout int) ([]string
+	return partitionMap
+}
+
+func seqEvenPartition(partitionMap *map[int][]int, numSubsets int, numNodes int) {
```
I don't think your map needs to be a pointer.

Go maps are pointers themselves, so it's not like your having to copy the whole map when you pass it as a parameter (citation needed, I'm not 100% sure but reasonably sure).

ZenGround0 replied:
You were right, nothing changes when I just make it a value

dgrisham commented:
@ZenGround0 @hsanjuan I meant to comment on this when you showed the original proposal, but what do you think about making the `subset_partition` section available a per-step option? The capabilities are certainly equivalent (between the global config and per-step), but the ability to do it per-step seems to add expressiveness. For example, if (for whatever reason) you wanted to run tests on 1/3 then 1/2 of nodes, you'd have to break into 2 * 3 = 6 subsets; this can get worse, e.g. for 1/6, then 1/7, then 1/8, you'd have to specify 6 * 7 * 8 = 256 subsets to get all of those options.

However, doing this per-step *only* could certainly get ugly/redundant, so (if we think this is a good idea at all), I propose two alternatives:

1. The global `config` always acts as the default, then there can be per-step overrides to change only for that step.
2. The option is not available in the `config`, but every time `subset_partition` is called in a step its configuration is used for all subsequent steps *until* another step changes `subset_partition`.

The 2nd option can get ugly in cases where the partitioning alternates every step, but seems a bit less confusing/more consistent than the 1st option since it doesn't mix global/local. Personally, I'm more inclined toward the 1st option, since it will probably clean up the amount of text in most tests that use the partitioning feature but still gives the option to override for expressiveness.

It's kind of hard to foresee what people might do with this, and therefore hard to choose the 'best' path, but this seems like it'd give a good balance in most cases.

ZenGround0 commented:
@dgrisham Thanks for the feedback.  I certainly understand where you are coming from.  Enforcing one partition is less expressive than allowing arbitrarily many partitions.  My current opinion is that we may not need to open up this functionality just yet, and perhaps not at all.  Looking at the current list of 8 tests on our todo list, each can be implemented with the DSL as it stands in this PR.  In addition the test that you mentioned can also be done in a straightforward way with this selection model using the Percent descriptor.  A test can specify that it wants to run 50% of nodes starting at node 1, for example, and then in the next step specify that it wants to run 33% of nodes starting at node 1, etc. 

My understanding is that the subset descriptor is valuable to have alongside the other descriptors because
1) it persists a partition, so we can randomly divide nodes into groups and keep referring to the same group
2) it makes it easier for the tester, especially in the even allocation group, to specify potentially annoying node index calculations concisely 

The Percent descriptor makes 2) less important, because while it can be less convenient I don't think it will be too bad for testers to calculate percents and starting indexes in cases like the test you describe above.  Because 1) is the only thing subset does that the other descriptors cannot do at all, and because 1) is a global property I think that the subset partition's natural place is in the config.

Because of all this my current plan is to wait to extend subset functionality until it becomes apparent that the extension would be helpful for tests we want to write.  However it is quite possible I am still overlooking something, in which case keep the feedback coming!

dgrisham commented:
@ZenGround0 Those are good points, especially about the Percent descriptor. I tried to think through what might be a minimal setup to provide for all of the options your current impl provides, and I only have one question after that: Are there obvious testing situations (or even ones that we plan on testing) that would use the Percent descriptor and *not* start at node 1? To me the main reason we'd want to do that is if we had some sort of heterogeneous purpose between sets of peers, e.g. peers 1 and 2 are treated as 'special' in a test and the rest (peers 3 ... n) are treated in a big batch, which might very well be a use case we currently have or want to cover.

However, if that's not something we want to cover, then it seems like the 1st alternative I mentioned would handle the Percent-descriptor-starting-at-node-1 case quite well. I think it really comes down to what we care more about: the ability to start at an arbitrary node and specify a percentage of nodes after that node, or to specify step-specific subset selections.

At this point, I don't know that the decision is terrible crucial and I think either option would work fairly well for a large chunk of forseeable cases. I mostly just wanted to conclude/document my thoughts on all of this, unless you have any further thoughts I'm happy with this implementation :)

dgrisham commented:
@ZenGround0 Also, I haven't look thoroughly through the code, but I glanced over it and it looks well-written (lots of functions, descriptive names, etc.). Nice coding style :)

dgrisham commented on main.go:
```diff
@@ -396,6 +511,187 @@ func runInPod(name string, cmdToRun string, env []string, timeout int) ([]string
+	} 
+	partitionMap := make(map[int][]int)
+
+	if config.SubsetPartition.Order == sequential {
```
Switch statements seem more suitable than if/elif in this block, since you're comparing the same variable repeatedly.

ZenGround0 commented:
@hsanjuan @dgrisham  I have addressed your feedback, exchanging many if/else blocks with switch statements, pre-validating step format and bubbling up errors.  The recent commit also refines the meta-testing structure, I added a file test_tests for all kubernetes-ipfs dsl tests and put the tests of the dsl feature in the selection_framework directory.  This commit also adds a suite of tests with predictable selection problems that should cause kubernetes-ipfs to fail before running steps in a subdirectory.

@dgrisham I feel you on the percentage start node not being too immediately useful.  Right now I think it's lightweight enough that it's acceptable to keep in, but if you feel strongly that it messes with simplicity I wouldn't be opposed to taking it out.

hsanjuan commented on main.go:
```diff
@@ -48,23 +57,96 @@ type Assertion struct {
+	Subset  *Subset  `yaml:"subset"`
+}
+
+// Range is
```
Still missing descriptions for Range, Percent, Subset, Config, SubsetPartition

hsanjuan commented on main.go:
```diff
@@ -168,34 +266,61 @@ func main() {
 			continue // skip handling the output or other assertions since it timed out.
 		}
+		if len(step.WriteToFile) != 0 {
+			f, err := os.OpenFile(step.WriteToFile, os.O_CREATE|os.O_RDWR|os.O_APPEND, 0664)
```
onliner https://golang.org/pkg/io/ioutil/#WriteFile

hsanjuan commented on test_tests/test_selection_framework.sh:
```diff
@@ -0,0 +1,20 @@
+#!/bin/bash
```
I think you can create `main_test.go` and load these yamls directly from Go and do the testing.

This will allow to get code coverage right away, which is useful to see if you left any code path out of the tests.

Are you familiar with how go testing works? https://golang.org/pkg/testing/ 

This will require refactoring the `main()` into more specialized functions that you can call from the tests. The main should probably just parse command-line arguments and call some sort of `Run(args)` which then does:

```
readTestFile()
validate()
RunTests()
PrintResults()
```

kind of thing.

ZenGround0 commented:
Another round of changes to address the most recent comments.  These two commits refactor main to more easily call it in parts, and add go framework testing in order to test the selection feature.  Please let me know what else remains to be fixed.

hsanjuan commented on main.go:
```diff
@@ -190,43 +212,56 @@ func main() {
+	var summary Summary
+	var subsetPartition map[int][]int
+
+	readTestFile(filePath, &test)
```
`test := readTestFile()`

Since Go functions allow to return stuff, let's use.

hsanjuan commented on main.go:
```diff
@@ -190,43 +212,56 @@ func main() {
+	var subsetPartition map[int][]int
+
+	readTestFile(filePath, &test)
+	validate(&test, &subsetPartition)
```
To my like, this would be more standard:

```
subsetPartition, err := partition(test.Config)
if !validate(test, subsetPartition) -> error
summary := RunTests(test, subsetPartition)
```

ZenGround0 replied:
I agree with moving partition out of validate.  I can see how validate returning a boolean could be more standard and clear in terms of reading the code, however I think it is important for validate to return a specific error message which lets the test writer know what they have done wrong to making fixing the test easier, so I am inclined to return an error from validate.

hsanjuan commented on main.go:
```diff
@@ -217,28 +217,35 @@ func main() {
 
-	readTestFile(filePath, &test)
-	validate(&test, &subsetPartition)
+	err := readTestFile(filePath, &test)
```
test := readTestFile(path)

ZenGround0 replied:
Ok, is the idea that the output will be nil if readTestFile fails?

hsanjuan commented on main.go:
```diff
@@ -217,28 +217,35 @@ func main() {
+	if err != nil {
+		fatal(err)
+	}
+	err = validate(&test, &subsetPartition)
```
```
subsetPartition := partition(test.Config)
if err := validate(test, subsetPartition); err != nil {...}
```

ZenGround0 commented:
@hsanjuan The last two commits should address the refactoring of main mentioned in the four comments of the latest round of feedback

dgrisham commented on main.go:
```diff
@@ -108,26 +212,60 @@ func main() {
 
-	var test Test
-	var summary Summary
+	rand.Seed(time.Now().UTC().UnixNano())
```
Looks like this line is always ran, maybe only want to do so when `RANDOM` partition is requested?

dgrisham commented on main.go:
```diff
@@ -108,26 +212,60 @@ func main() {
-
-	fileData, err := ioutil.ReadFile(filePath)
+	
+	var subsetPartition map[int][]int
```
This line meant to be deleted? Doesn't seem necessary given the `err2, subsetPartition := partition(test.Config)` below.

hsanjuan replied:
right

dgrisham commented on main.go:
```diff
@@ -108,26 +212,60 @@ func main() {
-	var test Test
-	var summary Summary
+	rand.Seed(time.Now().UTC().UnixNano())
+	err2, subsetPartition := partition(test.Config)
```
One more small thing -- reason for using `err`, then `err2`, then `err` again? Not sure what the best practice is here, I think I've seen `err` reused usually though.
